server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByFillWithValueFilterDataSet.java

```diff
@@ -0,0 +1,438 @@
+import org.apache.iotdb.db.qp.physical.crud.RawDataQueryPlan;
+import org.apache.iotdb.db.query.aggregation.AggregateResult;
+import org.apache.iotdb.db.query.context.QueryContext;
+import org.apache.iotdb.db.query.control.QueryResourceManager;
+import org.apache.iotdb.db.query.executor.fill.IFill;
+import org.apache.iotdb.db.query.executor.fill.LinearFill;
+import org.apache.iotdb.db.query.executor.fill.PreviousFill;
+import org.apache.iotdb.db.query.factory.AggregateResultFactory;
+import org.apache.iotdb.db.query.reader.series.IReaderByTimestamp;
+import org.apache.iotdb.db.query.reader.series.SeriesReaderByTimestamp;
+import org.apache.iotdb.db.query.timegenerator.ServerTimeGenerator;
+import org.apache.iotdb.tsfile.file.metadata.enums.TSDataType;
+import org.apache.iotdb.tsfile.read.expression.IExpression;
+import org.apache.iotdb.tsfile.read.expression.impl.BinaryExpression;
+import org.apache.iotdb.tsfile.read.expression.impl.SingleSeriesExpression;
+import org.apache.iotdb.tsfile.read.filter.GroupByFilter;
+import org.apache.iotdb.tsfile.read.filter.basic.BinaryFilter;
+import org.apache.iotdb.tsfile.read.filter.basic.Filter;
+import org.apache.iotdb.tsfile.read.query.timegenerator.TimeGenerator;
+import org.apache.iotdb.tsfile.utils.Pair;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class GroupByFillWithValueFilterDataSet extends GroupByFillEngineDataSet {
+
+  // the next query time range of each path
+  private List<TimeGenerator> timestampGenerators;
+  private List<TimeGenerator> extraPreviousGenerators;
+  private List<TimeGenerator> extraNextGenerators;
+
+  // data reader lists
+  private List<IReaderByTimestamp> allDataReaderList;
+  private List<IReaderByTimestamp> extraPreviousDataReaderList;
+  private List<IReaderByTimestamp> extraNextDataReaderList;
+
+  // cached timestamp for next group by partition
+  private long lastTimestamp;
+  private List<LinkedList<Long>> cachedTimestamps;
+
+  private final int timeStampFetchSize = IoTDBDescriptor.getInstance().getConfig().getBatchSize();
+
+  /** constructor. */
+  public GroupByFillWithValueFilterDataSet(
+      QueryContext context, GroupByTimeFillPlan groupByTimeFillPlan)
+      throws StorageEngineException, QueryProcessException {
+    super(context, groupByTimeFillPlan);
+
+    initPathGenerators(context, groupByTimeFillPlan);
+
+    initExtraGenerators(context, groupByTimeFillPlan);
+    if (extraPreviousGenerators != null) {
+      initExtraArrays(extraPreviousValues, extraPreviousTimes, true, extraPreviousGenerators);
+    }
+    if (extraNextGenerators != null) {
+      initExtraArrays(extraNextValues, extraNextTimes, false, extraNextGenerators);
+    }
+
+    initCachedTimesAndValues();
+  }
+
+  private void initPathGenerators(QueryContext context, GroupByTimeFillPlan groupByTimeFillPlan)
```

Comment:
Why is the method called `initPathGenerators`? I think it should be renamed to `initReadersByTimestamp()`

Reply:
done